### PR TITLE
[BEAM-7561] HdfsFileSystems.match can match a directory.

### PR DESCRIPTION
**Pull request by dmvk**
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by dmvk** [comment]
@timrobertson100 can you please take a look at this? I'm not hundred percent sure what the correct behavior should be

**Comment by timrobertson100** [comment]
Thanks @dmvk - will look tomorrow

File: sdks/java/io/hadoop-file-system/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java [hunk 1]
```diff
@@ -89,23 +89,21 @@
...
+          // recursive glob
           int index = spec.indexOf("**");
           metadata.addAll(
               matchRecursiveGlob(spec.substring(0, index + 1), spec.substring(index + 1)));
+        } else {
+          // normal glob
+          final FileStatus[] fileStatuses = fileSystem.globStatus(new Path(spec));
+          if (fileStatuses != null) {
+            for (FileStatus fileStatus : fileStatuses) {
+              metadata.add(toMetadata(fileStatus));
```
Comment (timrobertson100) [inline sdks/java/io/hadoop-file-system/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java]: Shouldn't the `if (fileStatus.isFile()) {...}` remain here?

Comment (timrobertson100) [reply sdks/java/io/hadoop-file-system/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java]: Maybe I'm confused with what you are trying to do... "HdfsFileSystems.match can match directory" could mean it does and shouldn't or that you want it to be able to match directories.

Can you open a Jira and document the issue and reference it in the PR please?

Comment (mareksimunek) [reply sdks/java/io/hadoop-file-system/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystem.java]: Expected behavior: `HadoopFileSystem.match("hdfs:///tmp/dir)"` returns `MatchResult` with `metadata` about directory (resourceId etc.)

Current behavior `HadoopFileSystem.match("hdfs:///tmp/dir)"` returns `MatchResult` with empty `metadata`

**Comment by timrobertson100** [comment]
Thanks @dmvk 
Looking at the specs for the `FileSystem` I think this looks correct behaviour.

It's a non-trivial change so needs a Jira and renamed commit following the guidelines. Is that something you can do please?

**Comment by dmvk** [comment]
@timrobertson100 thanks, will do; I've just wanted to have verified that this is an expected behavior first.

Thanks!

**Comment by timrobertson100** [comment]
Welcome 
I've worked only a little on FS stuff, but I've read around the various JDocs and agree this looks correct. In particular [this line](https://github.com/apache/beam/blob/dc02997418b472aeddb4fc6e24f3c3f94e454664/sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java#L52) which states `The user-provided {@code spec} might refer to files or directories`

**Comment by dmvk** [comment]
nit: I'm curious whether it would be possible to abstract `FileSystems` test cases as it is an universal interface, to make sure that all impls. have correct behavior.

**Comment by timrobertson100** [comment]
We should explore that. Can you please add the same test to LocalFileSystem to verify it behaves the same in the meantime?

**Comment by lukecwik** [comment]
+1 on a standard suite of tests. Also, there is a ResourceIdTester (
https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/fs/ResourceIdTester.java)
which is meant to validate ResourceId objects returned. We may want to
collapse the two together into one FileSystemTester.

**Comment by dmvk** [comment]
Jira issue for a follow-up https://issues.apache.org/jira/browse/BEAM-7591

**Comment by dmvk** [comment]
@timrobertson100 I've added the very same test case for LocalFileSystems. It would be great if we could merge this before a 2.14 release cut. There is also a Jira issue for followup that should unify test cases across impls. Thanks!! D.

**Comment by timrobertson100** [comment]
Thanks @dmvk

**Pull request merged**